tests/estimators/speech_recognition/test_tensorflow_lingvo.py

```diff
@@ -0,0 +1,329 @@
+                "norm_wer_errors",
+                "norm_wer_words",
+                "utt_id",
+            }
+        )
+
+    def test_predict(self, audio_data):
+        import tensorflow.compat.v1 as tf1
+
+        tf1.reset_default_graph()
+
+        test_input = audio_data
+
+        lingvo = LingvoAsr()
+        predictions = lingvo.predict(test_input, batch_size=2)
+        assert predictions.shape[0] == test_input.shape[0]
+        assert isinstance(predictions[0], np.str_)
+
+    def test_loss_gradient_tensor(self, audio_batch_padded):
+        import tensorflow.compat.v1 as tf1
+
+        tf1.reset_default_graph()
+
+        test_input, test_mask_frequency = audio_batch_padded
+        test_target_dummy = np.array(["DUMMY"] * test_input.shape[0])
+
+        lingvo = LingvoAsr()
+        feed_dict = {
+            lingvo._x_padded: test_input,
+            lingvo._y_target: test_target_dummy,
+            lingvo._mask_frequency: test_mask_frequency,
+        }
+        loss_gradient = lingvo._sess.run(lingvo._loss_gradient_op, feed_dict)
+        assert test_input.shape == loss_gradient.shape
+        assert loss_gradient.sum() == 0.0
+
+    def test_loss_gradient_per_batch(self, audio_data):
+        import tensorflow.compat.v1 as tf1
+
+        tf1.reset_default_graph()
+
+        test_input = audio_data
+        test_target = np.array(["This", "is", "a dummy", "a dummy"])
+
+        lingvo = LingvoAsr()
+
+        gradients = lingvo._loss_gradient_per_batch(test_input, test_target)
+        gradients_abs_sum = np.array([np.abs(g).sum() for g in gradients], dtype=object)
+
+        # test shape, equal inputs have equal gradients, non-zero inputs have non-zero gradient sums
+        assert [x.shape for x in test_input] == [g.shape for g in gradients]
+        assert_allclose(np.abs(gradients[2]).sum(), np.abs(gradients[3]).sum(), rtol=1e-01)
+        assert_array_equal(gradients_abs_sum > 0, [False, True, True, True])
+
+    def test_loss_gradient_per_sequence(self, audio_data):
+        import tensorflow.compat.v1 as tf1
+
+        tf1.reset_default_graph()
+
+        test_input = audio_data
+        test_target = np.array(["This", "is", "a dummy", "a dummy"])
+
+        lingvo = LingvoAsr()
+
+        gradients = lingvo._loss_gradient_per_sequence(test_input, test_target)
+        gradients_abs_sum = np.array([np.abs(g).sum() for g in gradients], dtype=object)
+
+        # test shape, equal inputs have equal gradients, non-zero inputs have non-zero gradient sums
+        assert [x.shape for x in test_input] == [g.shape for g in gradients]
+        assert_allclose(np.abs(gradients[2]).sum(), np.abs(gradients[3]).sum(), rtol=1e-01)
+        assert_array_equal(gradients_abs_sum > 0, [False, True, True, True])
+
+
+@pytest.mark.skipif(tf.__version__ != "2.1.0", reason="requires Tensorflow 2.1.0")
+@pytest.mark.skipMlFramework("pytorch", "mxnet", "kerastf", "non_dl_frameworks")
+class TestLingvoAsrLibriSpeechSamples:
```

Comment:
same thing as above with respect to the class

Reply:
See above.